### PR TITLE
Improve address normalization

### PR DESCRIPTION
**Pull request by lewandowski-stripe**
- Use the same address for connecting to destination and for ACL checks
- Support IPv6 addresses
- Remove regexp-based domain validation
- Enforce a standardized form of IP addresses and DNS names
- Add tests for `normalizeHost()`

### CODE REVIEW

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [inline pkg/smokescreen/smokescreen.go:670]: should this one also be wrapped?

```suggestion
		return "", denyError{pctx.Error}
```

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [inline pkg/smokescreen/smokescreen.go:939]: is this just because the regex failed on IPv6 addresses before?

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [inline pkg/smokescreen/smokescreen.go:894]: ACL checks will use the normalized host

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [inline pkg/smokescreen/smokescreen.go:900]: DNS resolution also uses the normalized host

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [inline pkg/smokescreen/smokescreen.go:565]: this is kind of interesting: we're now using the normalized host name instead of the one passed on the CONNECT request (which varies based on http clients). I think this makes sense though!

**Review by jjiang-stripe — CHANGES_REQUESTED** [review]
I think this looks good and is more consistent compared to the hacks I shoved in 😅

I left a couple minor comments! we should also deploy this internally to make sure there aren't any regressions (I can help with this!)

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:440]: Correct. Is the comment above sufficient or should I improve it?

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:447]: If we reached this code `hasPort()` returned true. So either it's a bare IPv6 address (already ruled out in the opening if statement) or the `hostPort` is indeed malformed (since it was supposed to contain a port but it does not).

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:459]: Yes. I think it's preferable to fail explicitly at this stage then assign a meaning to port 0.

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:467]: Not really. IP addresses can have multiple representations (e.g., `2001:DB8::` and `2001:db8::` represent the same address but they are different strings). We should ensure that we are consistent about which format we use.

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:670]: It probably should. It wasn't so I didn't touch it. ;)

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:939]: Yes. I hope to remove it in one of subsequent PRs as I hope to implement the IPv6 support. But for now I'd like to keep the current behaviour.

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:894]: Yes, WAI.

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:565]: We should use the same host that we used for ACL checking. This way, even if someone is able to workaround host normalization, they only should be able to access destinations they were already authorized for.

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go:900]: Correct!

File: pkg/smokescreen/smokescreen.go
Comment (jjiang-stripe) [reply pkg/smokescreen/smokescreen.go:440]: I think it makes sense! I'm mostly just papertrailing comments for myself so I remember that I understand what this is doing 😅

Comment (lewandowski-stripe) [reply pkg/smokescreen/smokescreen.go]: The code should be self explanatory so the places where you left comments are valuable to me as indicators where comments are missing or code should be made clearer. Thank you!

**Comment by lewandowski-stripe** [comment]
> I think this looks good and is more consistent compared to the hacks I shoved in 😅

Thank you! For what is worth, you addressed the problem quickly so I am glad that you have added these "hacks". ;)

> I left a couple minor comments! we should also deploy this internally to make sure there aren't any regressions (I can help with this!)

Thank you! That'd be very much appreciated!

**Review by jjiang-stripe — COMMENTED** [review]
this looks good! I'll make a TODO for us to test this out internally first before merging, and I'll come back to approve this once we do!

**Review by jjiang-stripe — DISMISSED** [review]
I deployed this internally to our QA environment and didn't see any DNS resolution failures or unexpected ACL denials, so I think this is good to go!

I also pulled this and did a little local repro for the square bracket CVEs we had _just to be super sure_ and they return denials/errors as expected.

**Comment by lewandowski-stripe** [comment]
r?

Thank you @jjiang-stripe!

File: pkg/smokescreen/smokescreen.go
Comment (lewandowski-stripe) [inline pkg/smokescreen/smokescreen.go]: Thank you @cds2-stripe for noticing the if-condition that'd never match!

**Review by cds2-stripe — APPROVED** [review]

**Pull request merged**
